stacks/api/bucket/src/bucket.schema.resolver.ts

```diff
@@ -1,30 +1,35 @@
 import {Injectable} from "@nestjs/common";
 import {BucketService, compile} from "@spica-server/bucket/services";
-import {Bucket} from "@spica-server/bucket/services/bucket";
+import {BucketPreferences} from "@spica-server/bucket/services/bucket";
 import {Validator} from "@spica-server/core/schema";
 import {ObjectId} from "@spica-server/database";
+import {map} from "rxjs/operators";
+import {from, combineLatest, Observable} from "rxjs";
 
 @Injectable()
 export class BucketSchemaResolver {
-  constructor(private bucketService: BucketService) {}
+  preferenceWatcher: Observable<BucketPreferences>;
+  constructor(private bucketService: BucketService) {
+    this.preferenceWatcher = this.bucketService.watchPreferences(true);
+  }
 
-  resolve(uri: string): Promise<Bucket | object> | undefined {
-    if (ObjectId.isValid(uri)) {
-      return this.bucketService.findOne({_id: new ObjectId(uri)}).then(async schema => {
-        const prefs = await this.bucketService.getPreferences();
-        if (schema) {
-          const jsonSchema = compile(schema, prefs);
-          jsonSchema.$id = uri;
-          jsonSchema.$schema = "http://spica.internal/bucket/schema";
-          jsonSchema.additionalProperties = false;
-          jsonSchema.properties._schedule = {
-            type: "string",
-            format: "date-time"
-          };
-        }
-        return schema || {type: true};
-      });
-    }
+  resolve(uri: string): Observable<object> {
+    return combineLatest(
+      this.preferenceWatcher,
+      from(this.bucketService.findOne({_id: new ObjectId(uri)}))
```

Comment:
How about we make a method in the bucket service that returns an observable which will emit the latest version of an arbitrary bucket whenever it changes? 

Therefore we could get rid of BucketCache once for all?

Reply:
If we make these suggested changes, we wouldn't need the BucketCache cause this class would both invalidate either of these (preference or bucket itself) changes.